### PR TITLE
PP-6189 Use gateway status if terminal instead of expiring

### PR DESCRIPTION
**Pull request by stephencdaly**
If a charge is picked up by the expiry sweep has a status that indicates
the gateway knows about it, then for ePDQ and Worldpay payments:

- we get the status of the charge with the gateway and map it to a
ChargeStatus.
- if the mapped status is terminal (CAPTURED, AUTHORISATION REJECTED)
then we don't expire and instead attempt to update the status of the
charge to match the gateway status.
- if the mapped status is not terminal, we attempt to cancel the charge
with the gateway and then expire it.

For other gateways, we do not check with the gateway and instead always
try to cancel with the gateway if the charge has been authorised
successfully.

## WHAT YOU DID
_A brief description of the pull request:_

## How to test

- How should it be reviewed? 
- What feedback do you need? 
- If manual testing is needed, give suggested testing steps.

### Logging

- [ ] only emit log lines at ERROR level which require immediate attention from a support engineer. These will trigger a zendesk alert.

### Documentation

- [ ] Updated README.md for any of the following ?

* Introduced any new environment variables / removed existing environment variable
* Added new API / updated existing API definition

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java [hunk 1]
```diff
@@ -161,38 +171,93 @@ private int expireChargesWithoutGateway(List<ChargeEntity> nonAuthSuccessCharges
...
+                                    kv(GATEWAY_ACCOUNT_ID, chargeEntity.getGatewayAccount().getId()),
+                                    kv(PROVIDER, chargeEntity.getGatewayAccount().getGatewayName()));
+
+                            try {
+                                chargeService.forceTransitionChargeState(chargeEntity, status);
+                                expireCancelFailed.getAndIncrement();
+                            } catch (InvalidForceStateTransitionException e) {
+                                logger.error(format("Cannot expire charge as it is in a terminal state of [%s] with the " +
+                                                "gateway provider. The charge's state could not be transitioned to [%s].",
+                                        status.getValue(), status.getValue()),
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java]: I don't think there's a need to log the same information `status.getValue()` twice here. A better log line might be `Cannot transition state from chargeEntity.getStatus() to status.getValue()`?

**Review by oswaldquek — CHANGES_REQUESTED** [review]
Looks good! One minor comment about logging.

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java (hunk 1 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java]: I pretty much lifted this logging from here https://github.com/alphagov/pay-connector/blob/b73bc6fbf61181cfcd11c2378a38c9484ebf183d/src/main/java/uk/gov/pay/connector/charge/service/ChargeCancelService.java#L103 
Should I change it there too?

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java]: I can do that as part of my up and coming PR for PP-6186!

Comment (stephencdaly) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java]: How about:

> Cannot expire charge as it is in a terminal state of [AUTHORISATION REJECTED] with the gateway provider and it is not possible to transition the charge into this state.

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java]: 👍 I think it'll also be good to log the charge's current state.

Comment (stephencdaly) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java]: updated

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java]: Sorry I meant 
```
Cannot expire charge as it is in a terminal state of [AUTHORISATION REJECTED] with the gateway provider and it is not possible to transition the charge into this state. Current state: [status.getValue()]
```

**Review by oswaldquek — DISMISSED** [review]

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java (hunk 1 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeExpiryService.java]: Updated

**Review by oswaldquek — APPROVED** [review]

**Pull request merged**